kv/index.go

```diff
@@ -0,0 +1,537 @@
+	// populateBatchSize configures the size of the batch used for insertion
+	populateBatchSize int
+	// canRead configures whether or not Walk accesses the index at all
+	// or skips the index altogether and returns nothing.
+	// This is used when you want to integrate only the write path before
+	// releasing the read path.
+	canRead bool
+}
+
+// IndexOption is a function which configures an index
+type IndexOption func(*Index)
+
+// WithIndexReadPathEnabled enables the read paths of the index (Walk)
+// This should be enabled once the index has been fully populated and
+// the Insert and Delete paths are correctly integrated.
+func WithIndexReadPathEnabled(i *Index) {
+	i.canRead = true
+}
+
+// WithIndexPopulateBatchSize configures the size of each batch
+// used when fully populating an index. (number of puts per tx)
+func WithIndexPopulateBatchSize(n int) IndexOption {
+	return func(i *Index) {
+		i.populateBatchSize = n
+	}
+}
+
+// IndexMapping is a type which configures and Index to map items
+// from a source bucket to an index bucket via a mapping known as
+// IndexSourceOn. This function is called on the values in the source
+// to derive the foreign key on which to index each item.
+type IndexMapping interface {
+	SourceBucket() []byte
+	IndexBucket() []byte
+	IndexSourceOn(value []byte) (foreignKey []byte, err error)
+}
+
+// IndexSourceOnFunc is a function which can be used to derive the foreign key
+// of a value in a source bucket.
+type IndexSourceOnFunc func([]byte) ([]byte, error)
+
+type indexMapping struct {
+	source []byte
+	index  []byte
+	fn     IndexSourceOnFunc
+}
+
+func (i indexMapping) SourceBucket() []byte { return i.source }
+
+func (i indexMapping) IndexBucket() []byte { return i.index }
+
+func (i indexMapping) IndexSourceOn(v []byte) ([]byte, error) {
+	return i.fn(v)
+}
+
+// NewIndexMapping creates an implementation of IndexMapping for the provided source bucket
+// to a destination index bucket.
+func NewIndexMapping(sourceBucket, indexBucket []byte, fn IndexSourceOnFunc) IndexMapping {
+	return indexMapping{
+		source: sourceBucket,
+		index:  indexBucket,
+		fn:     fn,
+	}
+}
+
+// NewIndex configures and returns a new *Index for a given index mapping.
+// By default the read path (Walk) is disabled. This is because the index needs to
+// be fully populated before depending upon the read path.
+// The read path can be enabled using WithIndexReadPathEnabled option.
+func NewIndex(mapping IndexMapping, opts ...IndexOption) *Index {
+	index := &Index{
+		IndexMapping:      mapping,
+		populateBatchSize: defaultPopulateBatchSize,
+	}
+
+	for _, opt := range opts {
+		opt(index)
+	}
+
+	return index
+}
+
+// IndexPopulatorStore is a store which also has a AutoPopulateIndex method
+// The method returns true when the store support auto population of index
+// on initialize.
+type IndexPopulatorStore interface {
+	Store
+	AutoPopulateIndex() bool
+}
+
+// Initialize creates the index bucket on the provided store
```

Comment:
note to self: document how given the store implements `AutoPopulateIndex()` and the method returns true, how ir automatically calls populate with the provided store.